src/TpetraLinearSystem.C

```diff
@@ -1474,64 +1492,50 @@ void TpetraLinearSystem::applyDirichletBCs(stk::mesh::FieldBase * solutionField,
+  NGPDoubleFieldType ngpBCValuesField = realm_.ngp_field_manager().get_field<double>(bcValuesField->mesh_meta_data_ordinal());
 
-    const stk::mesh::Bucket::size_type length   = b.size();
-    const double * solution = (double*)stk::mesh::field_data(*solutionField, *b.begin());
-    const double * bcValues = (double*)stk::mesh::field_data(*bcValuesField, *b.begin());
+  ngpSolutionField.sync_to_device();
+  ngpBCValuesField.sync_to_device();
 
-    Teuchos::ArrayView<const LocalOrdinal> indices;
-    Teuchos::ArrayView<const double> values;
-    std::vector<double> new_values;
+  auto entityToLID = entityToLID_;
+  const int maxOwnedRowId = maxOwnedRowId_;
+  const int maxSharedNotOwnedRowId = maxSharedNotOwnedRowId_;
+  auto ownedLocalMatrix = ownedLocalMatrix_;
+  auto sharedNotOwnedLocalMatrix = sharedNotOwnedLocalMatrix_;
+  auto ownedLocalRhs = ownedLocalRhs_;
+  auto sharedNotOwnedLocalRhs = sharedNotOwnedLocalRhs_;
 
-    for (stk::mesh::Bucket::size_type k = 0 ; k < length ; ++k ) {
-      const stk::mesh::Entity entity = b[k];
-      const stk::mesh::EntityId naluId = *stk::mesh::field_data(*realm_.naluGlobalId_, entity);
-      const LocalOrdinal localIdOffset = lookup_myLID(myLIDs_, naluId, "applyDirichletBCs");
+  nalu_ngp::run_entity_algorithm(
+    "TpetraLinSys::applyDirichletBCs", ngpMesh, stk::topology::NODE_RANK, selector,
+    KOKKOS_LAMBDA(const MeshIndex& meshIdx)
+    {
+      stk::mesh::Entity entity = (*meshIdx.bucket)[meshIdx.bucketOrd];
+      const LocalOrdinal localIdOffset = entityToLID[entity.local_offset()];
+      const bool useOwned = localIdOffset < maxOwnedRowId;
+      const LinSys::LocalMatrix& local_matrix = useOwned ? ownedLocalMatrix : sharedNotOwnedLocalMatrix;
+      const LinSys::LocalVector& localRhs = useOwned ? ownedLocalRhs : sharedNotOwnedLocalRhs;
+      const double diagonalValue = useOwned ? 1.0 : 0.0;
 
       for(unsigned d=beginPos; d < endPos; ++d) {
         const LocalOrdinal localId = localIdOffset + d;
-        const bool useOwned = localId < maxOwnedRowId_;
-        const LocalOrdinal actualLocalId = useOwned ? localId : localId - maxOwnedRowId_;
-        Teuchos::RCP<LinSys::Matrix> matrix = useOwned ? ownedMatrix_ : sharedNotOwnedMatrix_;
-        const LinSys::LocalMatrix& local_matrix = useOwned ? ownedLocalMatrix_ : sharedNotOwnedLocalMatrix_;
-
-        if(localId > maxSharedNotOwnedRowId_) {
-          std::cerr << "localId > maxSharedNotOwnedRowId_:: localId= " << localId << " maxSharedNotOwnedRowId_= " << maxSharedNotOwnedRowId_ << " useOwned = " << (localId < maxOwnedRowId_ ) << std::endl;
-          throw std::runtime_error("logic error: localId > maxSharedNotOwnedRowId_");
-        }
+        const LocalOrdinal actualLocalId = useOwned ? localId : localId - maxOwnedRowId;
 
-        // Adjust the LHS
+        NGP_ThrowRequire(localId <= maxSharedNotOwnedRowId);
```

Comment:
Should this be `NGP_ThrowAssert`? Wanted to understand the performance implications.

Reply:
Yes NGP_ThrowAssert would probably be better here. I'll add a commit and do that.